[somfytahoma] New channel on the gateway to execute scenes

lolodomo opened the pull request:
Signed-off-by: Laurent Garnier <lg.hc@free.fr>

lolodomo commented:
@octa22 : this is the implementation of action groups (scenes) through a simple channel on the bridge. For compatibility reasons, I did not remove thing stuff already implemented for action groups. Any user will then have the choice, either using the new channel on the bridge or create a thing for each action group he likes as before.

octa22 commented on bundles/org.openhab.binding.somfytahoma/src/main/resources/OH-INF/thing/bridge.xml:
```diff
@@ -9,6 +9,10 @@
 		<label>Somfy Tahoma Bridge</label>
 		<description>Somfy Tahoma bridge enabling communication with Somfy devices</description>
 
+		<channels>
+			<channel id="scenes" typeId="scenes"></channel>
+		</channels>
+
```
I still think a better place for this channel is Gateway, not Bridge. The Bridge thing is responsible for the cloud communication and should stay as is, without any channels. If something aggregates the devices in the scenarios, it is the Gateway.

lolodomo replied:
Ok, why not. I will move the channel. It makes the thing just a little more difficult.

lolodomo replied:
Done

octa22 commented on bundles/org.openhab.binding.somfytahoma/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -389,4 +389,10 @@
 		<state readOnly="true"/>
 	</channel-type>
+
+	<channel-type id="scenes">
+		<item-type>String</item-type>
+		<label>Scenes</label>
+		<description>The scenes defined in the cloud portal</description>
+	</channel-type>
```
Tahoma users knows ```scenarios``` and Tahoma internally uses ```actionsgroups```. Please do not introduce ```scenes``` and please rename the channel to ```scenarios```.

lolodomo replied:
I just used scene as I have the feeling that scene is the correct english word for what is a "scenario" in French.

I just found that:: "As nouns the difference between scene and scenario is that scene is the location of an event that attracts attention while scenario is an outline of the plot of a dramatic or literary work. "
So yes, scenario is the right word to use I think (now).

octa22 replied:
good to know it :-) So the Tahoma guys have translated it right

lolodomo replied:
Done

octa22 reviewed (changes requested):
I think having the scenarios as channels is a very interesting idea. But one universal channel seems to me to be very far from great user experience. I am missing any documentation in the readme, how the users would trigger the scenarios from a rule for instance? Do they need to know the internal OID of the actiongroup?

lolodomo commented:
> I think having the scenarios as channels is a very interesting idea. But one universal channel seems to me to be very far from great user experience.

Having to handle one thing for each action group looks far from great to me, just to select a command amongst a list of commands. That's why I introduced this command channel.

> I am missing any documentation in the readme, how the users would trigger the scenarios from a rule for instance? Do they need to know the internal OIDvof the actiongroup?

Good point, I forgot the documentation. And yes the users would have to use the internal OID of the action group in rules.
In UI, the action group label is automatically displayed to the user.

I already did that in other bindings like the hue binding for example, I added a console command to list the scenerios with their idea. I will do that.

lolodomo commented:
I will have to wait for the merge of PR #10355

octa22 reviewed (approved):
LGTM

octa22 reviewed (changes requested):
Please update the documentation

lolodomo commented:
Yes, remain to do to cover all your remarks:

* documentation
* console command

lolodomo commented:
@octa22 : ready for a final review.

lolodomo commented:
@octa22 : can you please finally validate ?

octa22 reviewed (approved):
LGTM. Thanks

lolodomo commented:
@openhab/add-ons-maintainers : please review & merge.

fwolter commented on bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaGatewayHandler.java:
```diff
@@ -52,21 +65,47 @@ public void initializeThing(@Nullable ThingStatus bridgeStatus) {
+            SomfyTahomaBridgeHandler handler = getBridgeHandler();
+            if (handler != null && command instanceof StringType) {
+                handler.executeActionGroup(command.toString());
+                updateState(channelUID, new StringType(command.toString()));
```
What is the reason for setting the Channel with the updated value manually?

lolodomo replied:
Because the channel state was not set if I did not add this line.
I tried only with MainUI. Could it be a setup problem with my item or a UI bug ?
I will try again, including with Basic UI.

fwolter replied:
Did you try to set the auto update policy to `recommended`? https://www.openhab.org/docs/developer/bindings/thing-xml.html#auto-update-policies

lolodomo replied:
Why the default policy would not work?

fwolter replied:
I'm not aware of the details of the auto update algorithm, but I think the default setting expects the Channel to be updated when the Thing is polled the next time. This will never happen in your case, so you can adjust the auto update policy to tell the framework to update the Channel immediately. Then, you can remove the `updateState()` and the model will also tell one the intented behavior of the Channel.

lolodomo replied:
I agree that it is the good approach. Tested and working.

lolodomo commented:
Sorry, I introduced a bad character in one of my commit in my email address. I believe it explains the DCO error. I tried to follow the instructions to fix it but without success.

fwolter commented:
Never mind, we can force the DCO to pass.

lolodomo commented:
@fwolter : build succeeded ;)
